libyul/backends/evm/OptimizedEVMCodeTransform.cpp

```diff
@@ -0,0 +1,527 @@
+			std::visit(util::GenericVisitor{
+				[&](LiteralSlot const& _literal)
+				{
+					m_assembly.setSourceLocation(locationOf(_literal));
+					m_assembly.appendConstant(_literal.value);
+				},
+				[&](FunctionReturnLabelSlot const&)
+				{
+					yulAssert(false, "Cannot produce function return label.");
+				},
+				[&](FunctionCallReturnLabelSlot const& _returnLabel)
+				{
+					if (!m_returnLabels.count(&_returnLabel.call.get()))
+						m_returnLabels[&_returnLabel.call.get()] = m_assembly.newLabelId();
+					m_assembly.setSourceLocation(locationOf(_returnLabel.call.get()));
+					m_assembly.appendLabelReference(m_returnLabels.at(&_returnLabel.call.get()));
+				},
+				[&](VariableSlot const& _variable)
+				{
+					if (m_currentFunctionInfo)
+						if (util::contains(m_currentFunctionInfo->returnVariables, _variable))
+						{
+							m_assembly.setSourceLocation(locationOf(_variable));
+							m_assembly.appendConstant(0);
+							return;
+						}
+					yulAssert(false, "Variable not found on stack.");
+				},
+				[&](TemporarySlot const&)
+				{
+					yulAssert(false, "Function call result requested, but not found on stack.");
+				},
+				[&](JunkSlot const&)
+				{
+					// Note: this will always be popped, so we can push anything.
+					// TODO: discuss if PC is in fact a good choice here.
+					// Advantages:
+					// - costs only 2 gas
+					// - deterministic value (in case it is in fact used due to some bug)
+					// - hard to exploit in case of a bug
+					// - distinctive, since it is not generated elsewhere
+					// Disadvantages:
+					// - static analysis might get confused until it realizes that these are always popped
+					// Alternatives:
+					// - any other opcode with cost 2
+					// - unless the stack is empty: DUP1
+					// - the constant 0
+					// Note: it might even make sense to introduce a specific assembly item for this, s.t.
+					//       the peephole optimizer can deal with this (e.g. POP PUSHJUNK can be removed).
+					m_assembly.appendInstruction(evmasm::Instruction::PC);
+				}
+			}, _slot);
+		},
+		// Pop callback.
+		[&]()
+		{
+			m_assembly.appendInstruction(evmasm::Instruction::POP);
+		}
+	);
+	yulAssert(m_assembly.stackHeight() == static_cast<int>(m_stack.size()), "");
+}
+
+void OptimizedEVMCodeTransform::operator()(CFG::BasicBlock const& _block)
+{
+	// Assert that this is the first visit of the block and mark as generated.
+	yulAssert(m_generated.insert(&_block).second, "");
+
+	auto const& blockInfo = m_stackLayout.blockInfos.at(&_block);
+
+	// Assert that the stack is valid for entering the block.
+	assertLayoutCompatibility(m_stack, blockInfo.entryLayout);
+	m_stack = blockInfo.entryLayout; // Might set some slots to junk, if not required by the block.
+	yulAssert(static_cast<int>(m_stack.size()) == m_assembly.stackHeight(), "");
+
+	// Emit jump label, if required.
+	if (auto label = util::valueOrNullptr(m_blockLabels, &_block))
+		m_assembly.appendLabel(*label);
```

Comment:
source location?

Reply:
That's what I mentioned in the channel yesterday: those are missing from the CFG so far and my plan was to add them there separately and then all the way through to here as small change after this PR.
I can also do it now, though.

Reply:
Have it in https://github.com/ethereum/solidity/pull/11922